### DISCUSSION TITLE
GT King Pro does not boot from SD card

### DISCUSSION THREAD

**Discussion by berndh1979** [johang/sd-card-images#18] [General]
Hi,

Just wrote the two GT King Pro 'Bullseye images' as suggested on the site to a sd card and put this into my Beelink GT King Pro. It does not boot from the sd card. It continues to boot from the emmc into the stock Android firmware provided by Beelink.

I tried to boot with a core-elec image flashed on the very same sd card (incl a dtb.img), and it did boot when I pushed the reset button for 5 seconds while powering up. So I thought I had to do that with the images provided as well. Unfortunately that did also not work.

Am I forgetting something? 

Thanks

**Comment by johang** [comment]
Le Potato board works fine. Different SoC but still Amlogic. Connect to the UART and show what is outputs during boot.

**Reply by berndh1979** [reply]
Hi, yeah thanks. I've ordered the RS232 cable and will try tonight.

**Comment by berndh1979** [comment]
This is my serial console output, it shows that it boots from the Android image on emmc:

[boot.txt](https://github.com/johang/sd-card-images/files/9813920/boot.txt)

**Reply by johang** [reply]
Do you have similar for the Bullseye SD card image?

**Reply by berndh1979** [reply]
This is with the bullseye sd card inserted into the sd card slot. No Reset pin pushed during power-on.

**Comment by berndh1979** [comment]
This is with the reset hold for 5sec during power on. It seems to start to init the sd card, but it can't find the recovery.bin on the sd card and still boots into Android.

[boot_with_reset_hold.txt](https://github.com/johang/sd-card-images/files/9814039/boot_with_reset_hold.txt)

**Reply by johang** [reply]
Can you show the same complete log but with core-elec on the SD card?

**Reply by berndh1979** [reply]
Sure, here you go:

[boot_coreelec.txt](https://github.com/johang/sd-card-images/files/9814431/boot_coreelec.txt)

**Reply by johang** [reply]
I think this happens because you have a valid bootloader on the board's eMMC. It has higher priority than SD card. But there seems to be a way to jump over to SD card if there is a `aml_autoscript` on the first partition. I can try to add this.

**Reply by johang** [reply]
Another method might be to erase the eMMC, but I guess it's more risky.

**Reply by berndh1979** [reply]
Indeed both the coreelec sd card image and another bullseye image from internet have both files in the FAT partition table, including for instance the dtb.imb, kernel etc. Also the aml_autoscript.

The sd card image on your site, only has a 48MB FAT empty partition and an ext4 full root-filesystem for the bullseye distro on partition 2 on the sd card. There is no booloader that points, nor configures that it needs to load files from partition 2 (ext4).

**Reply by berndh1979** [reply]
> Another method might be to erase the eMMC, but I guess it's more risky.

Yeah let's not do that (yet) :)

**Reply by johang** [reply]
The FAT partition is empty in most cases because the bootloader is written raw to the flash, before partition 1. The bootloader then loads boot.scr from partition 2, which in turn loads DTB and kernel from partition 2. It should be enough to add a aml_autoscript script that loads boot.scr from partition 2.

**Reply by johang** [reply]
Can you please try unzipping [aml_autoscript.zip](https://github.com/johang/sd-card-images/files/9823334/aml_autoscript.zip) and putting aml_autoscript in the FAT partition? I can't test this since my boards have empty eMMC.

**Reply by berndh1979** [reply]
Sure will do right away. On SD card or will USB boot work too?

**Reply by berndh1979** [reply]
Same boot to Android again... Sorry. Attached the bootlog again.

[boot_aml.txt](https://github.com/johang/sd-card-images/files/9823458/boot_aml.txt)

**Reply by johang** [reply]
Can you rename it to s905_autoscript and try again?

**Reply by berndh1979** [reply]
seems to do a lot more, but still booting Android :(

[boot_s905.txt](https://github.com/johang/sd-card-images/files/9823642/boot_s905.txt)

**Reply by johang** [reply]
`Unknown command 'ext4load' - try 'help'`

Yeah, that will never work. The bootloader on eMMC doesn't support loading from ext4.

**Reply by berndh1979** [reply]
Hmm, I see also other images that have an u-boot.ext on the FAT partition. This will 'update' the bootloader. Maybe we need more files on the boot partition.

**Reply by johang** [reply]
Do you have a link to a such image?

**Reply by berndh1979** [reply]
https://forum.bee-link.com/forum.php?mod=viewthread&tid=71846

As an example. They have multiple files on FAT. Unfortunately I have zero experience with bootloaders like this. It seems that during boot you can still configure the boot environments such that it will use a different bootloader and configure which device to boot from.

**Reply by berndh1979** [reply]
> Another method might be to erase the eMMC, but I guess it's more risky.

Do you have a gt king pro as well? If i know that erasing the emmc would do the trick it might still be an option as i wont be using the Android image anyway. I dont need the dual boot. 

Even better to actually boot linux from emmc, but one step at a time. :)

**Reply by berndh1979** [reply]
I managed to add a u-boot.ext (newer version taken from one of the other links) and a aml_autoscript on the FAT partition and now boot into the debian bullseye :) Yeah !!! still a couple of errors but it works. Perhaps some tweaking is needed to the aml_autoscript and it propably will work also from USB.

[boot_bullseye_working.txt](https://github.com/johang/sd-card-images/files/9832798/boot_bullseye_working.txt)

**Reply by johang** [reply]
Oh, that's crazy. So it's eMMC U-Boot -> s905_autoscript on FAT -> u-boot.ext on FAT -> /boot/boot.scr on ext4 -> kernel on ext4...

Can you share u-boot.ext and s905_autoscript? I think I can get this into the boot images in a proper way.

**Reply by berndh1979** [reply]
Yes indeed. I got the u-boot.ext and script from some other distro link, created by balbes150 and darkstar. You will find these names somewhere on the internet if you google them. They have created several distro already, but eventually dropped the work for support Beelink GT King. 

I've attached the zip files. All credits for these files go to them. 

[D.zip](https://github.com/johang/sd-card-images/files/9833005/D.zip)

**Reply by berndh1979** [reply]
> Oh, that's crazy. So it's eMMC U-Boot -> s905_autoscript on FAT -> u-boot.ext on FAT -> /boot/boot.scr on ext4 -> kernel on ext4...
> 
> Can you share u-boot.ext and s905_autoscript? I think I can get this into the boot images in a proper way.

I can now do a 'dd if=/dev/zero of=/dev/mmcblk1' to fully wipe the emmc, including the older bootloader on emmc. Perhaps this will already remove one part of the issue. With the emmc bootloader removed I guess we indeed have a simpler design for the sd card image. But with the original firmware on emmc we need to manipulate the old bootloader a bit to actually use the u-boot from FAT.

**Reply by johang** [reply]
This is still risky. I'm not sure how your particular device behaves without eMMC bootloader.

**Reply by berndh1979** [reply]
> This is still risky. I'm not sure how your particular device behaves without eMMC bootloader.

Haha indeed. I tried it.... bricked it.. :) Had to flash again with the USB_burning_tool from AMlogic. Now it boots again.

**Comment by berndh1979** [comment]
And this is the boot log after starten an sd card with coreelec and let it boot correct. Swap the sd card for the bullseye and then power off -> on again. It throws all kind of errors on incorrect fstype and searching for the kernel.img (I guess that is the coreelec kernel). 

In my boot image there is actually nothing. It is completely empty. Is that normal?

[boot_after_coreelec.txt](https://github.com/johang/sd-card-images/files/9814173/boot_after_coreelec.txt)

**Comment by berndh1979** [comment]
Completely different question. Do you also have a Bullseye image that is a little less 'pure'. For some reason it does not keep the time and it always starts again from Sunday 7 aug 2022. ????

Perhaps to do with the kernel itself....

**Reply by johang** [reply]
It will automatically set correct time if it can reach a NTP server. If not you can set your time once and run `/sbin/hwclock -w` to write to the RTC.
